R/functions-Spectrum.R

```diff
@@ -568,3 +587,462 @@ validSpectrum <- function(object) {
+        new_sp@intensity <- unlist(base::lapply(intsp, intensityFun),
+                                   use.names = FALSE)
+    } else {
+        new_sp@mz <- unlist(base::lapply(split(mzs, mz_groups), mzFun),
+                            use.names = FALSE)
+        new_sp@intensity <- unlist(base::lapply(split(ints, mz_groups),
+                                                intensityFun),
+                                   use.names = FALSE)
+    }
+    if (is.unsorted(new_sp@mz))
+        stop("m/z values of combined spectrum are not ordered")
+    new_sp@peaksCount <- length(new_sp@mz)
+    new_sp
+}
+
+#' Group a sorted `numeric` of m/z values from consecutive scans by ion assuming
+#' that the variation between m/z values for the same ion in consecutive scan
+#' is much lower than the difference between m/z values within one scan.
+#'
+#' @param x `numeric` with ordered and combined m/z values from consecutive
+#'     scans.
+#'
+#' @param mzd `numeric(1)` with the m/z difference below which m/z values are
+#'     grouped together. If not provided the `.estimate_mz_scattering` function
+#'     is used to estimate it.
+#' 
+#' @return `integer` of same length than `x` grouping m/z values.
+#'
+#' @noRd
+.group_mz_values <- function(x, mzd) {
+    mzdiff <- diff(x)
+    if (missing(mzd))
+        mzd <- .estimate_mz_scattering(mzdiff, TRUE)
+    ## Create a vector grouping values with difference in mz values being
+    ## smaller than mzd
+    nvals <- diff(c(0, which(!(mzdiff < mzd)), length(x)))
+    rep(1:length(nvals), nvals)
+}
+
+#' Estimate the extent of random scattering of m/z values of the same ion in
+#' consecutive scans. This bases on the assumption that the random scattering
+#' of m/z values is much smaller than the m/z resolution of the MS instrument.
+#'
+#' Assumes the data is in profile mode.
+#' 
+#' @param x Either values or differences between values.
+#'
+#' @param is_diff `logical(1)` indicating whether `x` are already differences.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.estimate_mz_scattering <- function(x, is_diff = FALSE) {
+    if (!is_diff)
+        x <- diff(x)
+    dens <- density(x, n = max(c(512, length(x) / 2)))
+    ## Find all turning points, i.e. where density is increasing
+    idxs <- which(diff(sign(diff(dens$y))) == 2)
+    if (length(idxs) == 0)
+        stop("Error estimating random scattering of m/z values in consecutive",
+             " scans: could not discriminate between expected m/z difference ",
+             " and random noise.")
+    dens$x[idxs[1]]
+}
+
+#' Estimate the m/z resolution (i.e. the average difference between m/z values)
+#' of a Spectrum.
+#'
+#' @param x `numeric` with the (sorted) m/z values of one spectrum.
+#'
+#' @return `numeric(1)` with the m/z resplution.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.estimate_mz_resolution <- function(x) {
+    d <- density(diff(x), n = max(c(512, length(x)/2)))
```

Comment:
I know I often ask for additional functions but wouldn't it be worth to use an own `.density` function to ensure that you always use the same values for `n` everywhere?
```r
.density <- function(x)stats::density(x, n=max(c(512L, length(x) / 2L))
```

Reply:
Makes sense. Thanks!